File: sdk/typescript/src/report.ts
export class InsertReport {
  inserted: Record<string, number> = {};
  skipped_fields: Record<string, Record<string, number>> = {};
  unsupported: Record<string, number> = {};

  addInserted(key: string, n = 1): void {
    this.inserted[key] = (this.inserted[key] ?? 0) + n;
  }

  addSkipped(typeName: string, field: string): void {
    const fields = (this.skipped_fields[typeName] ??= {});
    fields[field] = (fields[field] ?? 0) + 1;
  }

  addUnsupported(key: string, n: number): void {
    this.unsupported[key] = (this.unsupported[key] ?? 0) + n;
  }

  toJSON(): object {
    return {
      inserted: sortKeys(this.inserted),
      skipped_fields: Object.fromEntries(
        Object.keys(this.skipped_fields)
          .sort()
          .map((k) => [k, sortKeys(this.skipped_fields[k])]),
      ),
      unsupported: sortKeys(this.unsupported),
    };
  }
}

function sortKeys(o: Record<string, number>): Record<string, number> {
  return Object.fromEntries(Object.keys(o).sort().map((k) => [k, o[k]]));
}

File: sdk/julia/src/report.jl
"""
What an insert call wrote (`inserted`), and what it could not (`skipped_fields`,
`unsupported`). Zero counts are never stored.
"""
struct InsertReport
    inserted::Dict{String, Int}
    skipped_fields::Dict{String, Dict{String, Int}}
    unsupported::Dict{String, Int}
end

function InsertReport()
    return InsertReport(
        Dict{String, Int}(),
        Dict{String, Dict{String, Int}}(),
        Dict{String, Int}(),
    )
end

function add_inserted!(report::InsertReport, key::AbstractString, n::Int=1)
    report.inserted[key] = get(report.inserted, key, 0) + n
    return report
end

function add_skipped!(
    report::InsertReport,
    type_name::AbstractString,
    field::AbstractString,
)
    fields = get!(report.skipped_fields, type_name, Dict{String, Int}())
    fields[field] = get(fields, field, 0) + 1
    return report
end

function add_unsupported!(report::InsertReport, key::AbstractString, n::Int)
    report.unsupported[key] = get(report.unsupported, key, 0) + n
    return report
end

function report_dict(report::InsertReport)
    return Dict(
        "inserted" => report.inserted,
        "skipped_fields" => report.skipped_fields,
        "unsupported" => report.unsupported,
    )
end

Base.:(==)(a::InsertReport, b::InsertReport) = report_dict(a) == report_dict(b)
